onnxruntime/contrib_ops/cpu/triu.cc

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/framework/op_kernel.h"
+#include "core/util/math_cpuonly.h"
+#include "Eigen/src/Core/Map.h"
+#include "triu.h"
+#include <functional>
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    Triu,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", BuildKernelDefConstraints<float, double>()),
+    Triu);
+
+template <typename T>
+struct Triu::ComputeImpl {
+  void get_triu(const T* X_data, T* Y_data, int64_t matrix_h, int64_t matrix_w, int64_t k_val) const {
+    auto input_mat = ConstEigenMatrixMapRowMajor<T>(
+        X_data,
+        matrix_h,
+        matrix_w);
+    auto output_mat = EigenMatrixMapRowMajor<T>(
+        Y_data,
+        matrix_h,
+        matrix_w);
+
+    output_mat = input_mat;
+    for (int64_t i = -1 * matrix_h; i < k_val; i++) {
+      output_mat.diagonal(i).array() = static_cast<T>(0);
+    }
+  }
+
+  void operator()(const Tensor* X, Tensor* Y, int64_t k_val) const {
+    const auto& X_shape = X->Shape();
+    int64_t X_num_dims = static_cast<int64_t>(X_shape.NumDimensions());
+
+    const auto* X_data = X->Data<T>();
+    int64_t matrix_h = static_cast<int64_t>(X_shape[X_num_dims - 2]);
+    int64_t matrix_w = static_cast<int64_t>(X_shape[X_num_dims - 1]);
+
+    if (X_num_dims == 2) {
```

Comment:
Out of curiosity, I looked at what MSVC 2019 does for this: the compiler is inlining get_triu in one path, but not the other. The generated code will be smaller if you instead treat the X_num_dims case as batch_size=1 and just use the below loop. From the other suggestion, if you do x_shape.SizeToDimension(X_num_dims - 2) and X_num_dims=2, then you will get back a result of 1 which works for this.

Reply:
Thanks, this is very helpful.